### ISSUE TITLE
[Bugs] TestDeregistration does not process all incoming messages from AMF.

### ISSUE THREAD

**Issue by endrigoshi** [free5gc/free5gc#76]
**If you don't follow the issue template, your issue may be closed.**
**Please note this is an issue tracker, not a support forum. For general questions, please use [forum](https://forum.free5gc.org).**
**Also, please refer to [TS](https://github.com/free5gc/free5gc/wiki/Trouble_Shooting) and [forum](https://forum.free5gc.org) before bug reporting.**

## Describe the bug
In `TestDeregistration` part of the test, the procedure goes as follows:
- send NAS Deregistration Request (UE Originating)
- receive ngap UE Context Release Command
- send ngap UE Context Release Complete

However, after the `send NAS Deregistration Request (UE Originating)`, AMF replies with a `Deregistration accept (UE originating)`. This message is not handled in `TestDeregistration` which leads to errors when trying to register a second UE, without restarting the RAN-AMF connection.

## To Reproduce
- If you check Wireshark trace, you can see the message is sent from AMF, but not handled in the test file.
- Run by trying to register/deregister 2 UEs, without restarting the RAN-AMF connection.

## Expected behavior
The messages `Deregistration accept (UE originating)` sent from AMF should be handled in the test file.

## Environment (please complete the following information):
 - free5GC Version [e.g. v3.0.2]
 - OS: [e.g. Ubuntu 18.04 Server]
 - Kernel version [e.g. 5.0.0-23-generic]
- go version [e.g. 1.12.9 linux/amd64]
 - c compiler version (Option) [e.g. gcc version 7.5.0]

### Configuration File
Provide your configuration here.

If you don't know what to do, you can zip the `config` folder and upload the zip file here.

### PCAP File
Dump the packet and provide the pcap file here.

If you don't know what to do, you can use `sudo tcpdump -i any -w free5gc.pcap` before you run the bug reproduce. After that, upload the free5gc.pcap file here.

### Log File
Provide the program log file here.

If you do not know what to do, copy the log print on the screen and put it here.

## System architecture (Option)
Give the draft of your architecture. What's your scenario, bare metal use case or using k8s, etc.

## Walkthrough (Option)
What you have find or research for short.

**Comment by am677** [comment]
Hi, @free5gc-org 

I upgraded free5gc from v3.0.2 to v3.0.3.
It seems that free5gc v3.0.3 could not handle the Deregistration request (UE originating) message properly.
I confirmed that there was no difference of Deregistration requests between each version.
(see free5gc-pcap.zip)

I need to restart free5gc every time I reconnect the same UE after this upgrade.
I also hope this test scenario will be implemented ASAP.

**Comment by sumichaaan** [comment]
Hi, @am677 

Maybe I think this root cause is infinity loop in src/smf/producer/pdu_session.go:499.
I don't know if the code below is what the free5gc main developer intended, 
but I've confirmed that the De-registration sequence works for the time being.

```
diff --git a/producer/pdu_session.go b/producer/pdu_session.go
index a8e4a19..9f48269 100644
--- a/producer/pdu_session.go
+++ b/producer/pdu_session.go
@@ -499,7 +499,7 @@ func HandlePDUSessionSMContextRelease(rspChan chan smf_message.HandlerResponseMe
        for _, dataPath := range smContext.Tunnel.DataPathPool {

                dataPath.DeactivateTunnelAndPDR(smContext)
-               for curDataPathNode := dataPath.FirstDPNode; curDataPathNode != nil; curDataPathNode.Next() {
+               for curDataPathNode := dataPath.FirstDPNode; curDataPathNode != nil; curDataPathNode = curDataPathNode.Next() {
                        curUPFID, _ := curDataPathNode.GetUPFID()
                        if _, exist := deletedPFCPNode[curUPFID]; !exist {
                                seqNum = pfcp_message.SendPfcpSessionDeletionRequest(curDataPathNode.UPF.NodeID, smContext)
```

**Comment by am677** [comment]
Hi, @sumichaaan

Thank you for the information.
Applying your fix, I also confirmed that the re-registration sequence works without restarting Free5gc!

**Comment by free5gc-org** [comment]
Hi,
we has fixed this bug in v3.0.4
Can you upgrade free5gc to v3.0.4 and test if this problem is still happened?
